### DISCUSSION TITLE
Manual approval in pipeline step

### DISCUSSION THREAD

**Discussion by quinn** [concourse/concourse#8965] [Ideas]
Hi, my typical flow for CD has a "gate" between staging an prod. The pipeline is fully automated up to deploy on staging env, and a button to press to take the artifact deployed to staging and roll it out to prod. 

diagram courtesy of chatgpt:

```
+--------+       +---------+       +---------+       +----------+  
|        |       |         |       |         |       |          |  
| Commit +------>+  Build  +------>+ Staging +------>+ Approval |  
|        |       |         |       | Deploy  |       |  Gate(*) |  
+--------+       +---------+       +---------+       +----+-----+  
                                                          |  
                                             +------------+------------+  
                                             |                         |  
                                         (Approve)                 (Reject)  
                                             |                         |  
                                             v                         v  
                                        +----+----+              +-----+-----+  
                                        |         |              |           |  
                                        | Prod    |              | Release   |  
                                        | Deploy  |              | Discarded |  
                                        |         |              |           |  
                                        +---------+              +-----------+  
  
(*) Manual Step  
```

i tried searching keywords like "gate" "manual" "approval" but didn't find any relevant discussions.

**Comment by suhlig** [comment]
You need something that represents the approval, which you check for with a resource. I used the presence of a git tag for this in the past, but you can probably also use a file in a git repo, S3 bucket, etc. 

Your pipeline would then probably look similar to this (you have `Prod Deploy` where I have `build-it`):

<img width="1671" alt="image" src="https://github.com/concourse/concourse/assets/80507/e9ba2158-434f-4d80-a973-269660f98918">

Your `Prod Deploy` job would check for the presence of a tag (`v1.2.4-approved`) and kick off a new build. The `Release Discarded` could be implemented in the same way, e.g. looking for a tag `v1.2.4-rejected`. The `tag_filter` of the [git resource](https://github.com/concourse/git-resource) would be useful for this.

The connection between two jobs can be made with a `passed` statement:

```yaml
- name: build-it
  plan:
    - get: tagged-commits
      trigger: true
      passed: [ test-tag ]
```

The creation of the tag is independent of Concourse, but _could_ be implemented as a separate job that creates the appropriate tag.

**Reply by quinn** [reply]
OK, this makes sense! thank you.

**Comment by marco-m** [comment]
Worthwhile to consider is also https://github.com/Meshcloud/gate-resource, although I have not used it so I cannot give an experience report.
